### PR TITLE
Net standard 2.0 support for VS2015

### PR DESCRIPTION
**Pull request by jainaashish**
This PR is to support net standard 2.0 for VS2015. There is a new .net standard support extension for the .net frameworks so if this MSI is not installed and user tries to install a net standard 2.0 package into net461 project, then we will throw a error asking user to install this MSI in order to install this package.

Fixes https://github.com/NuGet/Home/issues/5325

@rrelyea

### CODE REVIEW

File: src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/VSMSBuildNuGetProjectSystem.cs [hunk 1]
```diff
@@ -929,5 +929,25 @@ public virtual IEnumerable<string> GetDirectories(string path)
                        select p.Name;
             });
         }
+
+        public bool IsNetStandardMSINeeded()
```
Comment (mishra14) [inline src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/VSMSBuildNuGetProjectSystem.cs:933]: please add some comments for public methods.

File: src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs [hunk 2]
```diff
@@ -217,6 +217,19 @@ public Task ExecuteScriptAsync(PackageIdentity identity, string packageInstallPa
             return Task.FromResult(0);
         }
 
+        public bool IsNetStandardMSINeeded()
```
Comment (mishra14) [inline src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs:220]: please add some comments for public methods.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs [hunk 3]
```diff
@@ -1676,12 +1676,23 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
...
                 {
                     nuGetProjectContext.OriginalPackagesConfig =
                         msbuildProject.PackagesConfigNuGetProject?.GetPackagesConfig();
+
+                    needsNetStandard20MSI = msbuildProject.IsNetStandardMSINeeded();
+                }
+
+                var DoesProjectTargetNet461ORGreater = false;
+                NuGetFramework targetFramework;
+                if (nuGetProject.TryGetMetadata<NuGetFramework>(NuGetProjectMetadataKeys.TargetFramework, out targetFramework))
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:1693]: nit: targetFramework can be declared as part of the function call.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs [hunk 4]
```diff
@@ -1761,6 +1772,21 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
                                 // use the version exactly as specified in the nuspec file
                                 var packageIdentity = downloadPackageResult.PackageReader.GetIdentity();
 
+                                // Check to see if the hack to install make net461/net462/net47 compatibile with netstandard20 or greater
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs]: nit: Check to see if the hack to make net461/net462/net47 compatibile with netstandard20 or greater

File: src/NuGet.Core/NuGet.PackageManagement/Utility/NetStandardCompatibilityUtil.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace NuGet.PackageManagement
+{
+    internal static class NetStandardCompatibilityUtil
+    {
+        internal static bool DoesProjectTargetsNEt461ORGreater(NuGetFramework currentProjectFramework)
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.PackageManagement/Utility/NetStandardCompatibilityUtil.cs]: typo: DoesProjectTargetsNet461ORGreater or DoesProjectTargetsNET461ORGreater

File: src/NuGet.Core/NuGet.PackageManagement/Utility/NetStandardCompatibilityUtil.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            if (currentProjectFramework == null)
+            {
+                throw new ArgumentNullException(nameof(currentProjectFramework));
+            }
+
+            if (currentProjectFramework.IsDesktop()
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.PackageManagement/Utility/NetStandardCompatibilityUtil.cs:18]: [IsDesktop ](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Frameworks/FrameworkExtensions.cs#L21)also returns true for `AspNet ` and `DNX`. Maybe we should check for `NetFramework `explicitly?

File: src/NuGet.Core/NuGet.PackageManagement/Utility/NetStandardCompatibilityUtil.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+            return false;
+        }
+
+        internal static bool IsNearestFrameworkNetStandard20OrGreater(NuGetFramework currentProjectFramework,
+            IEnumerable<NuGetFramework> supportedFrameworks)
+        {
+            // we look at target frameworks supported by the package and determine if the nearest framework is netstandard2.0,
+            // then we do need to install the compatibility package. We only do it once for the whole list of actions -
+            // hence the !needsNetstandard20Assets condition.
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.PackageManagement/Utility/NetStandardCompatibilityUtil.cs]: Where is this condition checked?

**Review by mishra14 — CHANGES_REQUESTED** [review]
Some comments/clarifications.

File: src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/VSMSBuildNuGetProjectSystem.cs (hunk 1 above)
Comment (jainaashish) [reply src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/VSMSBuildNuGetProjectSystem.cs:933]: As discussed offline, this is not a public public method, and for our understanding, name is self explanatory. I can add a comment down line to specify when we think MSI is installed.

File: src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs (hunk 2 above)
Comment (jainaashish) [reply src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs:220]: Explained on another comment

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs (hunk 3 above)
Comment (jainaashish) [reply src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:1693]: yeah, but this is more readable.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs (hunk 4 above)
Comment (jainaashish) [reply src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs]: corrected

File: src/NuGet.Core/NuGet.PackageManagement/Utility/NetStandardCompatibilityUtil.cs (hunk 6 above)
Comment (jainaashish) [reply src/NuGet.Core/NuGet.PackageManagement/Utility/NetStandardCompatibilityUtil.cs:18]: As discussed offline, this has been the same behavior for VS 2017 as well so we just back ported it. I think we shouldn't change anything here until we decides to explicitly not supporting ASPNet or DNX. And if we ever decides it then it should be changed for VS 2017 and not for VS 2015.

File: src/NuGet.Core/NuGet.PackageManagement/Utility/NetStandardCompatibilityUtil.cs (hunk 7 above)
Comment (jainaashish) [reply src/NuGet.Core/NuGet.PackageManagement/Utility/NetStandardCompatibilityUtil.cs]: copy-paste mistake, corrected the comment.

**Review by mishra14 — APPROVED** [review]
LGTM!

**Comment by BradBarnich** [comment]
The link doesn't work. https://aka.ms/netstandard-build-support-netfx goes to microsoft.com

**Pull request merged**